clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/ResizeAsgAtomicOperation.groovy

```diff
@@ -85,9 +88,44 @@ class ResizeAsgAtomicOperation implements AtomicOperation<Void> {
         .withDesiredCapacity(capacity.desired)
 
     autoScaling.updateAutoScalingGroup request
+
+
+    if (capacity.desired == 0) {
+      // there is an opportunity to expedite a resize to zero by explicitly terminating instances
+      // (server group _must not_ be attached to a load balancer)
+      terminateInstancesInAutoScalingGroup(
+        amazonClientProvider.getAmazonEC2(description.credentials, region, true),
+        describeAutoScalingGroups.getAutoScalingGroups().get(0) as AutoScalingGroup
+      )
+    }
+
     task.updateStatus PHASE, "Completed resize of ${asgName} in ${region}."
   }
 
+  static void terminateInstancesInAutoScalingGroup(AmazonEC2 amazonEC2, AutoScalingGroup autoScalingGroup) {
+    if (!autoScalingGroup.loadBalancerNames.isEmpty() || !autoScalingGroup.targetGroupARNs.isEmpty()) {
+      task.updateStatus(
+        PHASE,
+        "Skipping explicit instance termination, server group is attached to one or more load balancers"
+      )
+      return
+    }
+
+    def serverGroupName = autoScalingGroup.autoScalingGroupName
+    def instanceIds = autoScalingGroup.instances.instanceId
+
+    def terminatedCount = 0
+    instanceIds.collate(MAX_SIMULTANEOUS_TERMINATIONS).each {
+      try {
+        terminatedCount += it.size()
+        task.updateStatus PHASE, "Terminating ${terminatedCount} of ${instanceIds.size()} instances in ${serverGroupName}"
+        amazonEC2.terminateInstances(new TerminateInstancesRequest().withInstanceIds(it))
+      } catch (Exception e) {
+        task.updateStatus PHASE, "Unable to terminate instances, reason: '${e.message}'"
```

Comment:
I suspect a common exception we will see here is that we requested termination of an instance that has already been terminated. I suspect in that case we would then have left ~100 instances around.

I wonder if it would be worthwhile to do something like refresh the ASG, select 100 instances, terminate and then keep track of which have been terminated.

That said this is the same thing the destroy task is doing so... if the autoscaler has to eventually reap the remaining instances its probably not the worst thing